client/hwlib/src/collectors/hardware_info.rs

```diff
@@ -0,0 +1,324 @@
+/* Copyright 2024 Canonical Ltd.
+ * All rights reserved.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Written by:
+ *        Nadzeya Hutsko <nadzeya.hutsko@canonical.com>
+ */
+
+use anyhow::Result;
+use smbioslib;
+use time::macros::format_description;
+use time::Date;
+
+use crate::models::devices;
+
+pub fn load_smbios_data(
+    entry_filepath: &'static str,
+    table_filepath: &'static str,
+) -> Result<Option<smbioslib::SMBiosData>> {
+    let smbios_data = match table_load_from_device(entry_filepath, table_filepath) {
+        Ok(data) => Some(data),
+        Err(e) => {
+            eprintln!("Failed to load SMBIOS data: {}.", e);
+            None
+        }
+    };
+    Ok(smbios_data)
+}
+
+pub fn collect_bios_info(smbios_data: &smbioslib::SMBiosData) -> Result<devices::Bios> {
+    let bios_info = &smbios_data.collect::<smbioslib::SMBiosInformation>()[0];
+
+    let release_date_str = bios_info.release_date().to_string();
+    let release_date_format = format_description!("[month]/[day]/[year]");
+    let release_date_parsed = Date::parse(&release_date_str, &release_date_format)?;
+    let output_format = format_description!("[year]-[month]-[day]");
+    let release_date = release_date_parsed.format(&output_format)?;
+
+    let bios = devices::Bios {
+        firmware_revision: match (
+            bios_info.system_bios_major_release(),
+            bios_info.system_bios_minor_release(),
+        ) {
+            (Some(major), Some(minor)) => Some(format!("{}.{}", major, minor)),
+            _ => None,
+        },
+        release_date: Some(release_date),
+        revision: Some(bios_info.version().to_string()),
+        vendor: bios_info.vendor().to_string(),
+        version: bios_info.version().to_string(),
+    };
+    Ok(bios)
+}
+
+/// Retrieve CPU information from SMBios
+pub fn collect_processor_info_smbios(
+    smbios_data: &smbioslib::SMBiosData,
+    max_cpu_frequency_filepath: &'static str,
+) -> Result<devices::Processor> {
+    let processor_info = &smbios_data.collect::<smbioslib::SMBiosProcessorInformation>()[0];
```

Comment:
Many if these functions immediately call `.collect` on their passed `smbios_data`, to avoid duplicate work, would it be cleaner to take a `&smbioslib::SMBiosProcessorInformation` instead and do the collecting outside? Generally, with these kinds of cost, it can be nice to leave it up to the user of these functions (another example, consider the following)

```rust
struct MyThing(String);

impl MyThing {
    fn new(inner: &str) -> Self {
        // Costly clone hidden! Also, a clone is forced even if `inner` was
        // specifically created to be a parameter of this function, possibly
        // causing an unnecessary extra allocation+free.
        Self(inner.into())
    }

    fn new(inner: String) -> Self {
        // Any cloning done outside.
        Self(inner)
    }
}

Reply:
Updated in 72c7b97